File: longestJourney.java
import java.util.*;
import java.io.*;

public class Main {
    static List<Integer>[] adj = new ArrayList[100005];
    static boolean[] vis = new boolean[100005];
    static List<Integer> v = new ArrayList<>();
    
    public static void dfs(int s) {
        if (vis[s]) return;
        vis[s] = true;
        for (int i : adj[s]) dfs(i);
        v.add(s);
    }
    
    public static void solve() {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        
        for (int i = 0; i < 100005; i++) {
            adj[i] = new ArrayList<>();
        }
        
        for (int i = 0; i < m; i++) {
            int a = sc.nextInt();
            int b = sc.nextInt();
            adj[b].add(a);
        }
        
        for (int i = 1; i <= n; i++) {
            if (!vis[i]) dfs(i);
        }
        
        List<Integer> t = new ArrayList<>();
        int i = 0, f = 0;
        while (i < v.size()) {
            if (f == 0) {
                if (v.get(i) == 1) {
                    f = 1;
                    t.add(1);
                }
            } else {
                t.add(v.get(i));
            }
            i++;
        }
        
        int[] len = new int[100005];
        int[] par = new int[100005];
        len[1] = 1;
        par[1] = 0;
        
        for (int x : t) {
            for (int j : adj[x]) {
                if (len[x] < len[j] + 1 && len[j] > 0) {
                    len[x] = len[j] + 1;
                    par[x] = j;
                }
            }
        }
        
        i = n;
        v = new ArrayList<>();
        while (i != 0) {
            v.add(i);
            i = par[i];
        }
        Collections.reverse(v);
        
        if (v.size() < 2) {
            System.out.println("IMPOSSIBLE");
            return;
        }
        System.out.println(v.size());
        for (int x : v) {
            System.out.print(x + " ");
        }
    }
    
    public static void main(String[] args) {
        solve();
    }
}

File: longestJourney.py
def dfs(s):
    if vis[s]:
        return
    vis[s] = True
    for i in adj[s]:
        dfs(i)
    v.append(s)

def solve():
    n, m = map(int, input().split())
    global adj, vis, v
    adj = [[] for _ in range(100005)]
    vis = [False] * 100005
    v = []
    
    for _ in range(m):
        a, b = map(int, input().split())
        adj[b].append(a)
    
    for i in range(1, n + 1):
        if not vis[i]:
            dfs(i)
    
    t = []
    i, f = 0, 0
    while i < len(v):
        if f == 0:
            if v[i] == 1:
                f = 1
                t.append(1)
        else:
            t.append(v[i])
        i += 1
    
    len_arr = [0] * 100005
    par = [0] * 100005
    len_arr[1] = 1
    par[1] = 0
    
    for i in t:
        for j in adj[i]:
            if len_arr[i] < len_arr[j] + 1 and len_arr[j]:
                len_arr[i] = len_arr[j] + 1
                par[i] = j
    
    i = n
    v = []
    while i != 0:
        v.append(i)
        i = par[i]
    v.reverse()
    
    if len(v) < 2:
        print("IMPOSSIBLE")
        return
    print(len(v))
    print(" ".join(map(str, v)))

if __name__ == "__main__":
    solve()
